### PR TITLE
fix(sync) fix duplicate messages in sync mode

### PR DESCRIPTION
**Pull request by outsinre**
With sync mode and burst concurrency, message is duplicated to Kafka.

In the following data, we can see count 7, 9 and 14 are all duplicated!

```
{"body":"foo=bar&count=4&ts=1666339225.362","body_args":{"foo":"bar","ts":"1666339225.362","count":"4"}}
{"body":"foo=bar&count=10&ts=1666339225.362","body_args":{"foo":"bar","ts":"1666339225.362","count":"10"}}
{"body":"foo=bar&count=1&ts=1666339225.362","body_args":{"foo":"bar","ts":"1666339225.362","count":"1"}}
{"body":"foo=bar&count=2&ts=1666339225.362","body_args":{"foo":"bar","ts":"1666339225.362","count":"2"}}
{"body":"foo=bar&count=8&ts=1666339225.362","body_args":{"foo":"bar","ts":"1666339225.362","count":"8"}}
{"body":"foo=bar&count=6&ts=1666339225.362","body_args":{"foo":"bar","ts":"1666339225.362","count":"6"}}
{"body":"foo=bar&count=3&ts=1666339225.362","body_args":{"foo":"bar","ts":"1666339225.362","count":"3"}}
{"body":"foo=bar&count=5&ts=1666339225.362","body_args":{"foo":"bar","ts":"1666339225.362","count":"5"}}
{"body":"foo=bar&count=7&ts=1666339225.362","body_args":{"ts":"1666339225.362","foo":"bar","count":"7"}}
{"body":"foo=bar&count=9&ts=1666339225.362","body_args":{"ts":"1666339225.362","foo":"bar","count":"9"}}
{"body":"foo=bar&count=9&ts=1666339225.362","body_args":{"ts":"1666339225.362","foo":"bar","count":"9"}}
{"body":"foo=bar&count=7&ts=1666339225.362","body_args":{"ts":"1666339225.362","foo":"bar","count":"7"}}
{"body":"foo=bar&count=19&ts=1666339225.362","body_args":{"foo":"bar","ts":"1666339225.362","count":"19"}}
{"body":"foo=bar&count=20&ts=1666339225.362","body_args":{"foo":"bar","ts":"1666339225.362","count":"20"}}
{"body":"foo=bar&count=14&ts=1666339225.362","body_args":{"ts":"1666339225.362","foo":"bar","count":"14"}}
{"body":"foo=bar&count=12&ts=1666339225.362","body_args":{"ts":"1666339225.362","foo":"bar","count":"12"}}
{"body":"foo=bar&count=14&ts=1666339225.362","body_args":{"ts":"1666339225.362","foo":"bar","count":"14"}}
{"body":"foo=bar&count=13&ts=1666339225.362","body_args":{"foo":"bar","ts":"1666339225.362","count":"13"}}
{"body":"foo=bar&count=11&ts=1666339225.362","body_args":{"foo":"bar","ts":"1666339225.362","count":"11"}}
```

### CODE REVIEW

**Comment by outsinre** [comment]
Before fix:

```
root@test:/usr/local/lib/lua/lua-resty-kafka# busted -t test spec/producer_spec.lua
[==========] Running tests from scanned files.
[----------] Global test environment setup.
[----------] Running tests from spec/producer_spec.lua
[ RUN      ] spec/producer_spec.lua @ 24: Test producers:  avoid duplicate messages in sync mode #test
2022/11/08 06:16:42 [error] 358#0: *2 lua user thread aborted: runtime error: unknown reason
stack traceback:
coroutine 0:
        [C]: in function 'throw'
        /usr/local/share/lua/5.1/busted/core.lua:149: in function 'error'
        /usr/local/share/lua/5.1/luassert/assert.lua:51: in function 'is_nil'
        spec/producer_spec.lua:33: in function <spec/producer_spec.lua:28>
coroutine 1:
        [C]: in function 'wait'
        spec/producer_spec.lua:44: in function <spec/producer_spec.lua:24>
        [C]: in function 'xpcall'
        /usr/local/share/lua/5.1/busted/core.lua:178: in function 'safe'
        /usr/local/share/lua/5.1/busted/init.lua:40: in function 'executor'
        /usr/local/share/lua/5.1/busted/core.lua:314: in function </usr/local/share/lua/5.1/busted/core.lua:314>
        [C]: in function 'xpcall'
        /usr/local/share/lua/5.1/busted/core.lua:178: in function 'safe'
        /usr/local/share/lua/5.1/busted/core.lua:314: in function 'execute'
        /usr/local/share/lua/5.1/busted/block.lua:155: in function 'execute'
        /usr/local/share/lua/5.1/busted/init.lua:15: in function 'executor'
        /usr/local/share/lua/5.1/busted/core.lua:314: in function </usr/local/share/lua/5.1/busted/core.lua:314>
        [C]: in function 'xpcall'
        /usr/local/share/lua/5.1/busted/core.lua:178: in function 'safe'
        /usr/local/share/lua/5.1/busted/core.lua:314: in function 'execute'
        /usr/local/share/lua/5.1/busted/block.lua:155: in function 'execute'
        /usr/local/share/lua/5.1/busted/init.lua:7: in function 'executor'
        /usr/local/share/lua/5.1/busted/core.lua:314: in function </usr/local/share/lua/5.1/busted/core.lua:314>
        [C]: in function 'xpcall'
        /usr/local/share/lua/5.1/busted/core.lua:178: in function 'safe'
        /usr/local/share/lua/5.1/busted/core.lua:314: in function 'execute'
        /usr/local/share/lua/5.1/busted/execute.lua:58: in function 'execute'
        ..., context: ngx.timer
spec/producer_spec.lua:47: attempt to perform arithmetic on a nil value

stack traceback:
        spec/producer_spec.lua:47: in function <spec/producer_spec.lua:24>

[  ERROR   ] spec/producer_spec.lua @ 24: Test producers:  avoid duplicate messages in sync mode #test (106.07 ms)
[----------] 1 test from spec/producer_spec.lua (106.60 ms total)

[----------] Global test environment teardown.
[==========] 1 test from 1 test file ran. (106.69 ms total)
[  PASSED  ] 0 tests.
[  ERROR   ] 1 error, listed below:
[  ERROR   ] spec/producer_spec.lua @ 24: Test producers:  avoid duplicate messages in sync mode #test

 1 ERROR

```

**Comment by outsinre** [comment]
After fix:

```
root@test:/usr/local/lib/lua/lua-resty-kafka# busted -t test spec/producer_spec.lua
[==========] Running tests from scanned files.
[----------] Global test environment setup.
[----------] Running tests from spec/producer_spec.lua
[ RUN      ] spec/producer_spec.lua @ 24: Test producers:  avoid duplicate messages in sync mode #test
[       OK ] spec/producer_spec.lua @ 24: Test producers:  avoid duplicate messages in sync mode #test (4.53 ms)
[----------] 1 test from spec/producer_spec.lua (5.14 ms total)

[----------] Global test environment teardown.
[==========] 1 test from 1 test file ran. (5.22 ms total)
[  PASSED  ] 1 test.
```

**Comment by outsinre** [comment]
Blocks https://github.com/Kong/lua-resty-kafka/pull/24

File: dev/docker-compose.yaml [hunk 1]
```diff
@@ -21,31 +21,25 @@ services:
     image: confluentinc/cp-kafka:6.2.0
     networks:
       kafka:
+    depends_on:
+    - broker
     command: >
      /bin/bash -c "
-        while ! nc -z broker 9092;
```
Comment (jschmid1) [inline dev/docker-compose.yaml:26]: I added this hack as the `depends` token waits on service "readyness". Defining this term is up to the service and must not mean that the service is ready to accept requests.

If this bash block doesn't cause any issues I'd be inclined to keep it in as it guards against flakyness of tests.

**Review by jschmid1 — APPROVED** [review]
Up to you how to handle https://github.com/Kong/lua-resty-kafka/pull/23/files#r1016389327. Besides that, good to go 👍🏻

File: dev/docker-compose.yaml (hunk 1 above)
Comment (outsinre) [reply dev/docker-compose.yaml:26]: I remove this hack as `while` may loop forever if the 9092 port is not ready, especially at local dev environment when I want do some individual code and test.

It's safe to remove the hack as I can see the `healtcheck` in *broker* container.

**Pull request merged**